source/common/http/alternate_protocols_cache_impl.cc

```diff
@@ -0,0 +1,51 @@
+#include "common/http/alternate_protocols_cache_impl.h"
+
+namespace Envoy {
+namespace Http {
+
+AlternateProtocolsCacheImpl::AlternateProtocolsCacheImpl(ThreadLocal::Instance& tls,
+                                                         TimeSource& time_source)
+    : time_source_(time_source), slot_(tls) {
+  slot_.set([](Event::Dispatcher& /*dispatcher*/) { return std::make_shared<State>(); });
```

Comment:
Do you have a pointer towards some docs on this thread-local storage magic?

Reply:
No, but if you find some, I'd appreciate it! The interface is here:

include/envoy/thread_local/thread_local.h

Here's my sketch of how it works. Thread local storage has an allocator (ThreadLocal::Instance) which allocates pieces of storage (slots) which stores different objects for each thread. So when a slot is created, it needs to be "set" with a factory (lambda) for creating a fresh object of the right type each time a new thread accesses the slot.

But this is all what I think I understand. I'm not positive.